internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,532 @@
+	// The algorithm for producing an public key tweaked BIP-340 adaptor
+	// signature is as follows:
+	// This deviates from the original algorithm in steps 11 and 12.
+	//
+	// G = curve generator
+	// n = curve order
+	// d = private key
+	// m = message
+	// a = input randomness
+	// r, s = signature
+	//
+	// 1. d' = int(d)
+	// 2. Fail if m is not 32 bytes
+	// 3. Fail if d = 0 or d >= n
+	// 4. P = d'*G
+	// 5. Negate d if P.y is odd
+	// 6. t = bytes(d) xor tagged_hash("BIP0340/aux", t || bytes(P) || m)
+	// 7. rand = tagged_hash("BIP0340/nonce", a)
+	// 8. k' = int(rand) mod n
+	// 9. Fail if k' = 0
+	// 10. R = 'k*G
+	// 11. Check if R + T is odd. If it is, we need to try again with a new nonce.
+	// 12. e = tagged_hash("BIP0340/challenge", bytes(R + T) || bytes(P) || m) mod n
+	// 13. sig = bytes(R) || bytes((k + e*d)) mod n
+	// 14. If Verify(bytes(P), m, sig) fails, abort.
+	// 15. return sig.
+
+	// Step 1.
+	//
+	// d' = int(d)
+	var privKeyScalar btcec.ModNScalar
+	privKeyScalar.Set(&privKey.Key)
+
+	// Step 2.
+	//
+	// Fail if m is not 32 bytes
+	if len(hash) != scalarSize {
+		str := fmt.Sprintf("wrong size for message hash (got %v, want %v)",
+			len(hash), scalarSize)
+		return nil, errors.New(str)
+	}
+
+	// Step 3.
+	//
+	// Fail if d = 0 or d >= n
+	if privKeyScalar.IsZero() {
+		str := "private key is zero"
+		return nil, errors.New(str)
+	}
+
+	// Step 4.
+	//
+	// P = 'd*G
+	pub := privKey.PubKey()
+
+	// Step 5.
+	//
+	// Negate d if P.y is odd.
+	pubKeyBytes := pub.SerializeCompressed()
+	if pubKeyBytes[0] == secp.PubKeyFormatCompressedOdd {
+		privKeyScalar.Negate()
+	}
+
+	var privKeyBytes [scalarSize]byte
+	privKeyScalar.PutBytes(&privKeyBytes)
+	defer zeroArray(&privKeyBytes)
+	for iteration := uint32(0); ; iteration++ {
+		// Step 6-9.
+		//
+		// Use RFC6979 to generate a deterministic nonce k in [1, n-1]
+		// parameterized by the private key, message being signed, extra data
+		// that identifies the scheme, and an iteration count
+		k := btcec.NonceRFC6979(
+			privKeyBytes[:], hash, rfc6979ExtraDataV0[:], nil, iteration,
+		)
+
+		// Steps 10-15.
+		sig, err := schnorrEncryptedSign(&privKeyScalar, k, hash, pub, T)
+		k.Zero()
+		if err != nil {
+			// Try again with a new nonce.
+			continue
+		}
+
+		return sig, nil
+	}
+}
+
+func parseSig(sig *schnorr.Signature) (s *secp256k1.ModNScalar, r *secp256k1.FieldVal) {
+	sigB := sig.Serialize()
+	r = new(secp256k1.FieldVal)
+	r.SetByteSlice(sigB[0:32])
+	s = new(secp256k1.ModNScalar)
+	s.SetByteSlice(sigB[32:64])
+	return
+}
+
+// PrivateKeyTweakedAdaptorSig creates a private key tweaked adaptor signature.
+// This is created by a party which knows the hidden value.
+func PrivateKeyTweakedAdaptorSig(sig *schnorr.Signature, pubKey *btcec.PublicKey, t *secp256k1.ModNScalar) *AdaptorSignature {
+	T := new(secp256k1.JacobianPoint)
+	secp256k1.ScalarBaseMultNonConst(t, T)
+
+	s, r := parseSig(sig)
+	tweakedS := new(secp256k1.ModNScalar).Add(s).Add(t)
```

Comment:
Not a big deal, but you could do `new(secp256k1.ModNScalar).Add2(s, t)` to avoid adding to zero (which isn't free because it's all cryptographically constant time code).